verification.md

```diff
@@ -11,24 +12,296 @@ module VERIFICATION
+                     | #asByteStackInWidthaux(Int, Int, Int, WordStack) [function]
+  rule #asByteStackInWidth(X, N) => #asByteStackInWidthaux(X, N -Int 1, N, .WordStack)
+  rule #asByteStackInWidthaux(X, I, N, WS) => #asByteStackInWidthaux(X, I -Int 1, N, nthbyteof(X, I, N) : WS) when I >Int 0
+  rule #asByteStackInWidthaux(X, 0, N, WS) => nthbyteof(X, 0, N) : WS
+```
+
+Another type of byte-array manipulating operation is used to extract the function signature from the call data. The function signature is located in the first four bytes of the call data, but there is no atomic EVM instruction that can load only the four bytes, thus some kind of byte-twiddling operations are necessary.
+
+The extraction mechanism varies by language compilers. For example, in Viper, the first 32 bytes of the call data are loaded into the memory at the starting location 28 (i.e., in the memory range of 28 to 59), and the memory range of 0 to 31, which consists of 28 zero bytes and the four signature bytes, is loaded into the stack. In Solidity, however, the first 32 bytes of the call data are loaded into the stack, and the loaded word (i.e., a 256-bit integer) is divided by `2^(28*8)` (i.e., left-shifted by 28 bytes), followed by masked by 0xffffffff (i.e., 4 bytes of bit 1’s).
+
+The following lemmas essentially capture the signature extraction mechanisms. It reduces the reasoning efforts of the underlying theorem prover, factoring out the essence of the byte-twiddling operations.
+
+```{.k .uiuck}
+  // for viper
+  rule #padToWidth(N, #asByteStack(#asWord(WS))) => WS
+    requires noOverflow(WS) andBool N ==Int #sizeWordStack(WS)
+
+  // for solidity
+  rule #asWord(WS) /Int D => #asWord(#take(#sizeWordStack(WS) -Int log256Int(D), WS))
+    requires D %Int 256 ==Int 0 andBool D >=Int 0
+     andBool #sizeWordStack(WS) >=Int log256Int(D)
+     andBool noOverflow(WS)
+
+  // syntactic sugar
+  syntax Bool ::= noOverflow(WordStack)    [function]
+                | noOverflowAux(WordStack) [function]
+  rule noOverflow(WS) => #sizeWordStack(WS) <=Int 32 andBool noOverflowAux(WS)
+  rule noOverflowAux(W : WS) => 0 <=Int W andBool W <Int 256 andBool noOverflowAux(WS)
+  rule noOverflowAux(.WordStack) => true
+```
+
+### Abstraction for Hash
+
+We do not model the hash function as an injective function simply because it is not true due to the pigeonhole principle.
+
+Instead, we abstract it as an uninterpreted function that captures the possibility of the hash collision, even if the probability is very small.
+
+```
+  syntax Int ::= keccak ( WordStack ) [function, smtlib(smt_keccak)]
+```
+
+However, one can avoid reasoning about the potential collision by assuming all of the hashed values appearing in each execution trace are collision-free. This can be achieved by instantiating the injectivity property only for the terms appearing in the symbolic execution, in a way analogous to the universal quantifier instantiation.
```

Comment:
Not clear if you indeed did the above or not.  Where is it done?  How do you actually keep track of all the terms that appear in symbolic execution?  What is "universal quantifier instantiation"?

Reply:
By the quantifier instantiation, I mean something similar to the Z3 trigger for selectively instantiating the universally quantified axioms.

I implicitly did it for at least for ERC20. In ERC20, hash is only used for the hashed location in the storage, and they appear in the spec as the keys of `Map`, which implicitly means they are different. 

Of course, for a program that generate and manipulate the hashed values, we need to implement some new feature for that.

Reply:
Hm, this is tricky.  Please explain it as well as you can for now, including all the above.  We will then have to find a clean and general solution for it.  What is wrong with assuming them to be injective functions?

Reply:
Adding the universal injectivity axiom to z3 query is simply not sound.

Instantiating the injectivity axiom for each term created during the symbolic execution is already kinda done, since K considers non-functional symbols as constructor symbols. But when encoding z3 query, it doesn't add the universal injectivity axiom for them, which is still sound but not complete.

So, in some sense, the current symbolic execution engine is kind of doing what I explained in the above (except instantiating the injectivity axiom for each constructor term). But I'm not sure if it is still correct in terms of the new kore semantics and thus generating proof objects.